### DISCUSSION TITLE
Split keypad from modal editing

### DISCUSSION THREAD

**Discussion by lilactown** [meow-edit/meow#349] [Ideas]
I've gone through the meow tutorial and really enjoy this package's take on modal editing.

However, I'm really struggling with making it usable when combined with e.g. Doom. I've tried [doom-meow](https://github.com/meow-edit/doom-meow), and have responded to a couple of the issues over there. Ultimately, as it is now it's not usable for me in day to day editing as I rely on local leader keybinds for many things.

It seems to my naive eyes that most of these issues stem from conflicts with other packages desire to control the action taken on leader and localleader keys.

The keypad functionality seems like it's really useful if you're basing your config around meow, as it removes the need to pull in another package for its functionality, but if I want something more full featured like Doom or spacemacs' leader hey HUD, I can't _only_ use meow's modal editing. It would be lovely if I could do just that, so I thought I would bring it up here.

**Comment by eshrh** [comment]
I don't use doom so maybe this is a silly suggestion, but what about something like
```elisp
(meow-define-keys 'normal
  '("SPC" . ignore))
```

AFAIK, as long as you don't add anything to the leader map, this shouldn't (?) interfere with anything else. If it does, maybe we can work from there.

**Reply by lilactown** [reply]
You're right, and this is where I've landed after hours of experimenting.

The problem then moves over to Doom's side of the fence - actually, it's a problem with general.el. The detection of the leader key is dependent on the state that meow is in, and Doom/general.el doesn't have a way of hooking into that state.

I tried setting meow's `"SPC"` to `doom/leader`, which worked... EXCEPT for the localleader key 😩

**Comment by DogLooksGood** [comment]
You don't need doom-meow. You just disable evil, add meow. Then set
```
(setq meow-keypad-leader-dispatch doom-leader-map)
```

I don't use doom neither, but I think this should work.

**Comment by lilactown** [comment]
That seems to do a merge of meow's `C-c`, `C-<key>` leader key behavior and the doom leader map, in a weird way.
I could probably figure out how to disable the default meow leader key behavior but I've reached the limits of what I can spend on this for now. I also still do not see the localleader in the keymap.

FWIW I think this is a `general.el` and Doom problem. I can get most of the leader key stuff to work, whether I choose to use meow or whatever general.el does, EXCEPT for the localleader. The way general.el configures the localleader is totally opaque to me, and when I try and dig into general.el's source I get bogged down in so many caveats, old code, and code that interfaces just with `evil` in an implicit, global way.

When I try and intercept the `define-localleader-key!` macro like in doom-meow, I end up with every mode's localleader merged together. IDK how general.el handles keeping them separate and I've given up trying to figure it out. Maybe the next person will.

Rather than try and kludge meow and general.el together in a way that neither were meant to, I think I'll start building out a scratch config around meow.

The only thing I really miss when using meow's default keypad is that none of my prefixes have labels. I know it's silly, but TBH if I could do that I would just copy the bits of config I wanted from doom into my own config. I might try hydra next.

**Reply by DogLooksGood** [reply]
The builtin way to name a keymap is `defalias`.

**Comment by lyjdwh** [comment]
When I use spacemacs，`(setq meow-keypad-leader-dispatch spacemacs-cmds)` achieved a similar effect. However space + x/h/c  behave as   
same as meow x/h/c in normal state; space + m/g refer to M- / C-M-

and my which key become to this
![image](https://user-images.githubusercontent.com/14898811/206394895-f2d80bd7-96b0-4e19-bd8b-1a74ea58b312.png)

Is there any way to fix this for spacemacs that space x/h/c/m/g behave in spacemacs way? 

Thanks !

**Reply by DogLooksGood** [reply]
`meow-keypad` is a command bound to `SPC`. You can bind `SPC` to spacemacs leader key map.

**Reply by lyjdwh** [reply]
Thanks, It works!

**Comment by chriad** [comment]
I also use the Spacemacs distribution with vim editing style (evil). I was a little bit confused at first but two things clarified it for me (also mentioned in the comments):
1. Change editing style to `emacs` in your spacemacs config file. `meow` and `evil` in a sense exclude each other.
2. keybindings
The variable `dotspacemacs-emacs-leader-key` is `M-m` (For `vim` flavor it was "SPC"). So with `meow` you access the spacemacs bindings with `SPC m m`. For example `spacemacs/switch-buffer` was on `SPC b b` and is now on `SPC m m b b`.

Major mode leader key is a shortcut key which is the equivalent of pressing `<leader> m`. So instead of just "," now you have to do `SPC m m m`.

This all can probably be shortened a bit by redifining the spacemacs leader keys. 

I also did this
```elisp
   (meow-leader-define-key (cons "SPC" spacemacs-default-map))
```
Then I can use `SPC SPC` as with vim editing mode to access spacemacs top level bindings.

One thing that I do not know how to solve is that `which-key` shows all the prefix keys with "+prefix" (see screenshots in the comments above) instead of with the identifier usually shown in Spacemacs (for example "+format" etc.)
